From 4952d8deb5c1cd422b45339c7139c48d475032bb Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 03 Dec 2020 20:39:11 +0200
Subject: [PATCH] [SW-20141] gaudi: fix nic qmans disable

The gaudi_disable_nic_qmans() function was ported from Linux, but
Preboot takes registers (and bases) defines from specs, so the base is
a full 39 bit physical page address. So when the function adds an offset
to the mmNIC0_QM0_BASE value, it actually changes the lower 32 bits, but
overflows... This results in incorrect address access and stalls the
SoC.
Fix this by clearing the upper bits which correspond to the CFG base of
the chip.

Change-Id: I176b363cc5872abc27aaab3e6d62fec47ae06250
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
(cherry picked from commit 58753ac4565bcb8d338a0f52d953348600ed9119)
---

diff --git a/gaudi/reset.c b/gaudi/reset.c
index 46ebc6e..81020c2 100644
--- a/gaudi/reset.c
+++ b/gaudi/reset.c
@@ -186,10 +186,11 @@
 	u32 nic_offset = 0;
 	u32 nic_delta_between_qmans = mmNIC0_QM1_BASE - mmNIC0_QM0_BASE;
 	u32 nic_delta_between_nics = mmNIC1_QM0_BASE - mmNIC0_QM0_BASE;
+	u32 nic_base = mmNIC0_QM0_BASE & ~HL_CFG_BASE;
 	int nic_id;
 
 	for (nic_id = 0; nic_id < GAUDI_NIC_ENGS_NUM; nic_id++) {
-		regmap_writel(0, mmNIC0_QM0_BASE + nic_offset);
+		regmap_writel(0, nic_base + nic_offset);
 
 		nic_offset += nic_delta_between_qmans;
 		if (nic_id & 1) {
